paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.paimon.spark
+
+import org.apache.paimon.predicate.{Predicate, PredicateBuilder}
+import org.apache.paimon.table.{AppendOnlyFileStoreTable, Table}
+
+import org.apache.spark.sql.connector.read.{Scan, ScanBuilder, SupportsPushDownFilters, SupportsPushDownLimit, SupportsPushDownRequiredColumns}
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+
+class SparkScanBuilder(table: Table)
+  extends ScanBuilder
+  with SupportsPushDownFilters
+  with SupportsPushDownRequiredColumns
+  with SupportsPushDownLimit {
+
+  private var predicates: Option[Predicate] = None
+
+  private var pushed: Option[Array[Filter]] = None
+
+  private var projectedIndexes: Option[Array[Int]] = None
+
+  private var pushDownLimit: Option[Int] = None
+
+  override def build(): Scan = {
+    val readBuilder = table.newReadBuilder()
+
+    projectedIndexes.foreach(readBuilder.withProjection)
+    predicates.foreach(readBuilder.withFilter)
+    pushDownLimit.foreach(readBuilder.withLimit)
+
+    new SparkScan(table, readBuilder);
+  }
+
+  override def pushFilters(filters: Array[Filter]): Array[Filter] = {
+    val converter = new SparkFilterConverter(table.rowType)
+    val predicates: Array[Predicate] = filters.flatMap {
+      filter =>
+        try {
+          Some(converter.convert(filter))
+        } catch {
+          case _: UnsupportedOperationException =>
+            None
+        }
+    }
+    if (predicates.nonEmpty) {
+      this.predicates = Some(PredicateBuilder.and(predicates: _*))
+    }
+    this.pushed = Some(filters)
+    filters
+  }
+
+  override def pushedFilters(): Array[Filter] = {
+    pushed.getOrElse(Array.empty)
+  }
+
+  override def pruneColumns(requiredSchema: StructType): Unit = {
+    val pruneFields = requiredSchema.fieldNames
+    val fieldNames = table.rowType.getFieldNames
+    val projected = pruneFields.map(field => fieldNames.indexOf(field))
+    this.projectedIndexes = Some(projected)
+  }
+
+  override def pushLimit(limit: Int): Boolean = {
```

Comment:
Does Spark SQL will not push limit if there is filter?

Reply:
yes. SparkSQL can guarantee this.